src/Access/AccessControlManager.cpp

```diff
@@ -63,6 +63,7 @@ class AccessControlManager::ContextAccessCache
         if (x)
             return *x;
         auto res = std::shared_ptr<ContextAccess>(new ContextAccess(manager, params));
+	    res->subscribeForChanges(res);
```

Comment:
I didn't quite understood how it will fix the issue. I mean you moved `subscribeForChanges` out of the constructor `ContextAccess::ContextAccess()`, but how this is going to be helpful?

Reply:
I'm sorry I didn't explain clearly.

As smart pointer, [res] in `getContextAccess()` is held by [context_access_cache],which means[context_access_cache] mananges many ContextAccess objects
in access management cycle. [context_access_cache] is always existing but ContextAccess object could be created and relasesed. 

In the original version, the function the constructor ContextAccess() hold a [this] pointer
 in a anonymous functions, which would cause potential coredump when the ContextAccess object which [this] pointer point to is released because [this] pointer could be Unknown.

 In the version we change, we ensure [res] in anonymous functions can always be known, which can avoid the potential coredump risk.

Reply:
Yes, but when `ContextAccess` is destroyed, its destructor releases `subscription_for_user_change` so [this lambda function](https://github.com/ClickHouse/ClickHouse/blob/e13a90e641b8ab05e582ba9e3c2ce9d120a32ffe/src/Access/ContextAccess.cpp#L190) must not be called after the `~ContextAccess()` has been called.

Reply:
I looked at the coredump. It seems it shows another problem: perhaps [this lambda function](https://github.com/ClickHouse/ClickHouse/blob/e13a90e641b8ab05e582ba9e3c2ce9d120a32ffe/src/Access/ContextAccess.cpp#L190) was called while another thread was destroying ContextAccess. I mean probably something like the following could happen:
1) First thread was handling a GRANT query, so it started updating all related instances of `ContextAccess`.
2) Second thread at the same time decided to evict `ContextAccess` from the cache, so it started destroying it.
3) First thread didn't know that `ContextAccess` is being destroyed, so it called that lambda function and got the failure https://github.com/ClickHouse/ClickHouse/issues/33253

If my guess is right the problem can be fixed not by changing the constructor but by writing an explicit destructor for `ContextAccess` and calling `setUser(nullptr)` in there:

```
ContextAccess::~ContextAccess()
{
    std::lock_guard lock{mutex};
    setUser(nullptr);
}
```

Probably it would be very nice to have a test reproducing the issue so we could check that it's fixed for sure.

Reply:
I thought about the issue a bit more. It seems in order to get rid of the crash completely we need your change and my proposed change in the destructor and probably even weak pointers inside the lambda. So your PR is ok and I think I'll add a little to it.